File: packages/telemetry/src/evaluate/evaluate-stage-gates.ts
import type { GatePolicyInput, GateRunContext } from '@ff-promo/contracts';
import type { StageGateEvaluation } from '@ff-promo/contracts';
import type { PrometheusClient } from '../client/prometheus-client.js';
import { evaluateGatePolicy } from './evaluate-gate-policy.js';

export async function evaluateStageGates(
	client: PrometheusClient,
	policies: GatePolicyInput[],
	runContext: GateRunContext,
): Promise<StageGateEvaluation> {
	const results = [];
	for (const policy of policies) {
		results.push(await evaluateGatePolicy(client, policy, runContext));
	}

	const verdict = results.every((result) => result.verdict === 'pass')
		? 'pass'
		: 'fail';

	return { verdict, results };
}

File: kotlin/modules/telemetry/src/main/kotlin/com/ffpromo/telemetry/evaluate/EvaluateStageGates.kt
package com.ffpromo.telemetry.evaluate

import com.ffpromo.contracts.GatePolicyInput
import com.ffpromo.contracts.GateRunContext
import com.ffpromo.contracts.GateVerdictResult
import com.ffpromo.contracts.StageGateEvaluation
import com.ffpromo.telemetry.client.PrometheusClient

suspend fun evaluateStageGates(
    client: PrometheusClient,
    policies: List<GatePolicyInput>,
    runContext: GateRunContext,
): StageGateEvaluation {
    val results = policies.map { policy ->
        evaluateGatePolicy(client, policy, runContext)
    }

    val verdict = if (results.all { it.verdict == GateVerdictResult.pass }) {
        GateVerdictResult.pass
    } else {
        GateVerdictResult.fail
    }

    return StageGateEvaluation(verdict = verdict, results = results)
}
